adapters/repos/db/inverted/tracker/propertyIdTracker.go

```diff
@@ -0,0 +1,147 @@
+//                           _       _
+// __      _____  __ ___   ___  __ _| |_ ___
+// \ \ /\ / / _ \/ _` \ \ / / |/ _` | __/ _ \
+//  \ V  V /  __/ (_| |\ V /| | (_| | ||  __/
+//   \_/\_/ \___|\__,_| \_/ |_|\__,_|\__\___|
+//
+//  Copyright © 2016 - 2023 Weaviate B.V. All rights reserved.
+//
+//  CONTACT: hello@weaviate.io
+//
+
+package tracker
+
+import (
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"sync"
+)
+
+type JsonPropertyIdTracker struct {
+	path        string
+	LastId      uint64
+	PropertyIds map[string]uint64
+	sync.Mutex
+}
+
+func NewJsonPropertyIdTracker(path string) (*JsonPropertyIdTracker, error) {
+	t := &JsonPropertyIdTracker{
+		path:        path,
+		PropertyIds: make(map[string]uint64),
+		LastId:      1,
+	}
+
+	log.Printf("Loading property id tracker from %v\n", path)
+
+	// read the file into memory
+	bytes, err := os.ReadFile(path)
+	if err != nil {
+		if os.IsNotExist(err) {
+			t.Flush(false)
+			return t, nil
+		}
+		return nil, err
+	}
+
+	// Unmarshal the data
+
+	if err := json.Unmarshal(bytes, &t); err != nil {
+		return nil, err
+	}
+	t.path = path
+
+	// Leave zero undefined to catch undefined key lookups
+	if t.LastId == 0 {
+		t.LastId = 1
+	}
+
+	return t, nil
+}
+
+// Writes the current state of the tracker to disk.  (flushBackup = true) will only write the backup file
+func (t *JsonPropertyIdTracker) Flush(flushBackup bool) error {
+	if !flushBackup { // Write the backup file first
+		t.Flush(true)
+	}
+
+	t.Lock()
+	defer t.Unlock()
+
+	bytes, err := json.Marshal(t)
+	if err != nil {
+		return err
+	}
+
+	filename := t.path
+	if flushBackup {
+		filename = t.path + ".bak"
+	}
+
+	// Do a write+rename to avoid corrupting the file if we crash while writing
+	tempfile := filename + ".tmp"
+
+	err = os.WriteFile(tempfile, bytes, 0o666)
+	if err != nil {
+		return err
+	}
+
+	err = os.Rename(tempfile, filename)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Drop removes the tracker from disk
+func (t *JsonPropertyIdTracker) Drop() error {
```

Comment:
If the file is already deleted, the function has succeeded, so it is successful.

Resetting the tracker is up to the calling code.  It can always reset the tracker by creating a new one.  Accidentally resetting the tracker would corrupt the entire database, so we don't do it.